### PR TITLE
Continuous batching

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Draft continous batching for eager attention

TODO:
- test on gpu for the paged attention path
- better scheduling: sort request with keys being amount of blocks required, you only have to take the 1st sequence smaller than this
- better cache object: not having to create both the cache index and the fill index would earn some time
- profile + performance optimisation

### CODE REVIEW

File: src/transformers/generation/continuous_batching.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+                "cache_index": cache_index,
+                "fill_index": fill_index,
+                "cache":paged_attention_cache,
+            }
+            out = self.model.forward(
+                continous_batch, position_ids=position_ids,  **kwargs
+            ).last_hidden_state
+            logits = self.lm_head(out[:,current_batch.cumulative_seqlens_q, :])
+            # we don't sample for now :)
+            logits = torch.softmax(logits, dim=-1)
```
Comment (qgallouedec) [inline src/transformers/generation/continuous_batching.py:400]: do you need the softmax here? subsequent argmax will return the same index with or without it

Comment (ArthurZucker) [reply src/transformers/generation/continuous_batching.py:400]: For sampling yeah normally you do softmax
